src/mongo/db/pipeline/document_source_backup_cursor.cpp

```diff
@@ -0,0 +1,190 @@
+    code of portions of this program with the OpenSSL library under certain
+    conditions as described in each individual source file and distribute
+    linked combinations including the program with the OpenSSL library. You
+    must comply with the Server Side Public License in all respects for
+    all of the code used other than as permitted herein. If you modify file(s)
+    with this exception, you may extend this exception to your version of the
+    file(s), but you are not obligated to do so. If you do not wish to do so,
+    delete this exception statement from your version. If you delete this
+    exception statement from all source files in the program, then also delete
+    it in the license file.
+======= */
+
+#define MONGO_LOGV2_DEFAULT_COMPONENT ::mongo::logv2::LogComponent::kQuery
+
+#include "mongo/platform/basic.h"
+
+#include "mongo/db/pipeline/document_source_backup_cursor.h"
+
+#include "mongo/logv2/log.h"
+
+namespace mongo {
+
+namespace {
+constexpr StringData kDisableIncrementalBackup = "disableIncrementalBackup"_sd;
+constexpr StringData kIncrementalBackup = "incrementalBackup"_sd;
+constexpr StringData kBlockSize = "blockSize"_sd;
+constexpr StringData kThisBackupName = "thisBackupName"_sd;
+constexpr StringData kSrcBackupName = "srcBackupName"_sd;
+
+// We only link this file into mongod so this stage doesn't exist in mongos
+REGISTER_DOCUMENT_SOURCE(backupCursor,
+                         LiteParsedDocumentSourceDefault::parse,
+                         DocumentSourceBackupCursor::createFromBson);
+}  // namespace
+
+using boost::intrusive_ptr;
+
+const char* DocumentSourceBackupCursor::getSourceName() const {
+    return kStageName.rawData();
+}
+
+Value DocumentSourceBackupCursor::serialize(
+    boost::optional<ExplainOptions::Verbosity> explain) const {
+    return Value(Document{
+        {getSourceName(),
+         Document{
+             {kDisableIncrementalBackup,
+              _backupOptions.disableIncrementalBackup ? Value(true) : Value()},
+             {kIncrementalBackup, _backupOptions.incrementalBackup ? Value(true) : Value()},
+             {kBlockSize, Value(_backupOptions.blockSizeMB)},
+             {kThisBackupName,
+              _backupOptions.thisBackupName ? Value(*_backupOptions.thisBackupName) : Value()},
+             {kSrcBackupName,
+              _backupOptions.srcBackupName ? Value(*_backupOptions.srcBackupName) : Value()}}}});
+}
+
+DocumentSource::GetNextResult DocumentSourceBackupCursor::doGetNext() {
+    if (_backupCursorState.preamble) {
+        Document doc = _backupCursorState.preamble.get();
+        _backupCursorState.preamble = boost::none;
+
+        return doc;
+    }
+
+    if (_docIt != _backupInformation.end()) {
+        Document doc = {{"filename"_sd, _docIt->first},
+                        {"fileSize"_sd, static_cast<long long>(_docIt->second.fileSize)}};
+        ++_docIt;
+
+        return doc;
+    }
+
+    return GetNextResult::makeEOF();
+}
```

Comment:
The backup cursor usage example in the catalog/README.md mentions that at a getmore command needs to be sent as a 'heartbeat' at least once every 10 mins to keep the backup cursor alive.

https://github.com/mongodb/mongo/blob/master/src/mongo/db/catalog/README.md#copy-files-while-sending-heartbeats

10 mins is the default for cursorTimeoutMillis, maybe the agg pipeline $backupCursor stage's cursor is just inheriting that and we don't need to add anything. But let's check that we can send a getmore without having an error, can do backups where one file takes > 10 mins to copy.

Reply:
I tested getMore command command. It works without errors. The interesting part is "batchSize" parameter which exists in "getMore" and also in aggregation stage itself. I will test if "batchSize" is respected and will put update on that here.

Reply:
"batchSize" works as expected.